soc/xtensa/intel_adsp/common/include/cpu_init.h

```diff
@@ -101,4 +101,42 @@ static ALWAYS_INLINE void cpu_early_init(void)
 	XTENSA_WSR("VECBASE", reg);
 }
 
+/* NOTE: This struct will grow with all values that have to be stored for
+ * proper cpu restore after PG.
+ */
+struct core_state {
+	uint32_t a0;
+	uint32_t a1;
+	uint32_t vecbase;
+	uint32_t excsave2;
+	uint32_t excsave3;
+	uint32_t thread_ptr;
+	uint32_t intenable;
+	uint32_t bctl;
+};
+
+static ALWAYS_INLINE void _save_core_context(struct core_state *core_desc, uint32_t core_id)
+{
+	core_desc[core_id].vecbase = XTENSA_RSR("VECBASE");
+	core_desc[core_id].excsave2 = XTENSA_RSR("EXCSAVE2");
+	core_desc[core_id].excsave3 = XTENSA_RSR("EXCSAVE3");
+	core_desc[core_id].thread_ptr = XTENSA_RUR("THREADPTR");
+	__asm__ volatile("mov %0, a0" : "=r"(core_desc[core_id].a0));
+	__asm__ volatile("mov %0, a1" : "=r"(core_desc[core_id].a1));
```

Comment:
Yikes.  You're stashing the return address and stack pointer here, in a random function in the middle of some C processing that is going to **continue**, then at restore you push these values back into the registers and not the ones that were actually present in the code doing the suspend!?

I mean... I see that this is ALWAYS_INLINE, which means that this code isn't clearly 100% wrong, but... yikes.  All it takes is for someone to helpfully move this into a sub-function and everything will blow up in a way that almost no one will be able to successfully debug.

Can we please move the GPR saving to the point of suspend so that you restore to exactly the same state that suspended?

(Alternatively: have the suspend happen out of a known-restartable context like the idle thread, and just run a fixed function at startup that doesn't need to restore GPR state)

Reply:
@andyross  save_core_context is already in idle thread when powering off dsp core.
````
 idle thread
    pm_system_suspend
       pm_state_set ( SOFT_OFF)
          save_dsp_context
````
After restore_dsp_context, the idle thread will get out of pm_state_set and go on executing. Do I miss something ? thanks!

Reply:
This is not a random place and its not a function you can call from random place. It's executed at the end of cpu shutdown process and and we don't have much flexibility here when it comes to moving this code to other places.